clickhouse/columns/array.h

```
@@ -38,19 +53,209 @@ class ColumnArray : public Column {
+    explicit ColumnArrayT(ColumnArray&& array)
+        : ColumnArray(std::move(array))
+        , typed_nested_data_(this->GetData()->template AsStrict<NestedColumnType>())
+    {}
+
+    /** Create a ColumnArrayT from a ColumnArray, without copying data and offsets.
+     *  Ownership of column internals is transferred to returned object, original (argument) object
+     *  MUST NOT BE USED IN ANY WAY, it is only safe to dispose it.
+     *
+     *  Throws an exception of `col` is of wrong type, it is safe to use original col in this case.
+     *  This is a static method to make such conversion verbose.
+     */
+    static auto Wrap(ColumnArray&& col) {
+        if constexpr (std::is_base_of_v<ColumnArray, NestedColumnType> && !std::is_same_v<ColumnArray, NestedColumnType>) {
+            // assuming NestedColumnType is ArrayT specialization
+
+            auto result = std::make_shared<ColumnArrayT<NestedColumnType>>(NestedColumnType::Wrap(col.GetData()));
+            result->offsets_ = col.offsets_;
+
+            return result;
+        } else {
+            return std::make_shared<ColumnArrayT<NestedColumnType>>(std::move(col));
+        }
+    }
+
+    static auto Wrap(Column&& col) {
+        return Wrap(std::move(dynamic_cast<ColumnArray&&>(col)));
+    }
+
+    // Helper to simplify integration with other APIs
+    static auto Wrap(ColumnRef&& col) {
+        return Wrap(std::move(*col->AsStrict<ColumnArray>()));
+    }
+
+    class ArrayWrapper {
+        const std::shared_ptr<NestedColumnType> typed_nested_data_;
+        const size_t offset_;
+        const size_t size_;
+
+    public:
+        using ValueType = typename NestedColumnType::ValueType;
+
+        ArrayWrapper(std::shared_ptr<NestedColumnType> data, size_t offset = 0, size_t size = std::numeric_limits<size_t>::max())
+            : typed_nested_data_(data)
+            , offset_(offset)
+            , size_(std::min(typed_nested_data_->Size() - offset, size))
```

If `size` is bigger than `typed_nested_data_->Size() - offset`, I assume the client would expect to get `typed_nested_data_->Size() - offset` of size. If that's the case, then it's fine. Otherwise, maybe we should throw an exception.

---

well, we build `ArrayWrapper` ourselves, and IMO such situation would never happen.